cmd/insights/contributors.go

```diff
@@ -0,0 +1,309 @@
+		}(repoURL)
+	}
+	go func() {
+		waitGroup.Wait()
+		close(doneChan)
+	}()
+	select {
+	case <-doneChan:
+		return nil
+	case err := <-errorChan:
+		close(errorChan)
+		return err
+	}
+}
+
+type repositoryContributorInsights struct {
+	RepoID  int
+	RepoURL string
+	New     []string
+	Recent  []string
+	Churn   []string
+	Repeat  []string
+}
+
+func (rci *repositoryContributorInsights) RenderTable() {
+	tableRows := []table.Row{
+		{"Repository URL", rci.RepoURL},
+		{"New contributors", len(rci.New)},
+		{"Recent contributors", len(rci.Recent)},
+		{"Churn contributors", len(rci.Churn)},
+		{"Repeat contributors", len(rci.Repeat)},
+	}
+	repoTable := table.NewWriter()
+	repoTable.AppendRows(tableRows)
+	repoTable.SetOutputMirror(os.Stdout)
+	repoTable.SetStyle(table.StyleBold)
+	repoTable.Style().Options.SeparateRows = true
+	repoTable.Render()
+}
+
+func getRepositoryContributorInsightsAll(opts *Options, repoURL string) (*repositoryContributorInsights, error) {
+	repo, err := findRepository(opts.APIClient, repoURL)
+	if err != nil {
+		var errRepoNotFoundOrIndexed *ErrRepoNotFoundOrIndexed
+		if errors.As(err, &errRepoNotFoundOrIndexed) {
+			fmt.Println("ignoring repository issue:", errRepoNotFoundOrIndexed)
+			return nil, nil
+		}
+		return nil, fmt.Errorf("could not get contributor insights for repository %s: %w", repoURL, err)
+	}
+	query := fmt.Sprintf("repoIds=%d&range=%d", repo.ID, opts.Period)
+	repoContributorInsights := &repositoryContributorInsights{
+		RepoID:  repo.ID,
+		RepoURL: repoURL,
+	}
+	var (
+		errorChan = make(chan error)
+		doneChan  = make(chan bool)
+		waitGroup = new(sync.WaitGroup)
+	)
+	waitGroup.Add(4)
+	go func() {
+		defer waitGroup.Done()
+		response, err := getRepositoryContributorInsights(opts.APIClient, "new", query)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.New = append(repoContributorInsights.New, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := getRepositoryContributorInsights(opts.APIClient, "recent", query)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Recent = append(repoContributorInsights.Recent, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := getRepositoryContributorInsights(opts.APIClient, "churn", query)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Churn = append(repoContributorInsights.Churn, data.AuthorLogin)
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		response, err := getRepositoryContributorInsights(opts.APIClient, "repeat", query)
+		if err != nil {
+			errorChan <- err
+			return
+		}
+		for _, data := range response.Data {
+			repoContributorInsights.Repeat = append(repoContributorInsights.Repeat, data.AuthorLogin)
+		}
+	}()
+	go func() {
```

Comment:
Very cool concurrency going on here! have been looking into also doing this on #24 as well to parallelize the data collection of the contributors so that the TUI can get the initial model started faster. Some questions if you don't mind. For this part, could it be the case that this is a dead goroutine when an error is returned from one of the data collectors, like the select is blocking once the error is acquired from ```errorChan``` closes the channel and then returns nil and err but waitGroup is not done, so ```doneChan``` does not get close?

Also,  could this be a good case to use [errGroup](https://pkg.go.dev/golang.org/x/sync@v0.3.0/errgroup)? @jpmcb

Reply:
Hi @k1nho, thanks!
The `waitGroup` will always close, regardless if an error occurred or not. This is because the start of each go routine explicitly defers the call to `waitGroup.Done()`.
In my opinion, this is a suitable place to use `errGroup`. It just comes down to style/preference. Behind the hood, they're doing something similar if I'm not wrong.